### PR TITLE
Improve retrieval of volume details

### PR DESCRIPTION
**Pull request by psathyan**
Signed-off-by: Pragadeeswaran Sathyanarayanan <psathyan@redhat.com>

# Description

In this PR, improvements to volume create & attach methods have been carried out. 

[Logs](http://magna002.ceph.redhat.com/ceph-qe-logs/psathyan/openstack/volume.log)

### CODE REVIEW

File: mita/openstack.py [hunk 1]
```diff
@@ -310,63 +310,46 @@ def _wait_until_ip_is_known(self):
...
-            )
-            try:
-                _vol = self.driver_v2.create_volume(
-                    self.size_of_disk, "{}-vol-{}".format(self.node_name, item)
-                )
-                self.volumes.append(_vol)
-            except BaseException as be:  # noqa
-                logger.debug("Failed to create volume.")
-                logger.debug(be)
+            vol_name = f"{self.node_name}-vol-{item}"
```
Comment (ktdreyer) [inline mita/openstack.py:313]: It's much clearer when you've moved `vol_name` to its own assignment. Thanks.

File: mita/openstack.py [hunk 2]
```diff
@@ -310,63 +310,46 @@ def _wait_until_ip_is_known(self):
...
-        ok_states = ["creating"]  # it's ok to wait if the volume is in this
+    def _wait_until_volume_available(self, volume: StorageVolume) -> bool:
+        """Wait until the state of the StorageVolume is available."""
         tries = 0
-        if maybe_in_use:
-            ok_states.append("in_use")
-        logger.info("Volume: %s is in state: %s", volume.name, volume.state)
-
-        while volume.state in ok_states:
+        logger.debug("%s waiting for volume to become available.", volume.name)
```
Comment (ktdreyer) [inline mita/openstack.py:335]: Let's rewrite this string as `Waiting for %s volume to become available`

File: mita/openstack.py [hunk 3]
```diff
@@ -310,63 +310,46 @@ def _wait_until_ip_is_known(self):
...
+                return True
+
+            if "error" in volume.state.lower():
+                logger.error("%s state is %s", volume.name, volume.state)
                 break
-            if volume.state == "notfound":
-                logger.error("no volume was found for: %s", volume.name)
+
+            if tries > 10:
+                logger.error("Maximum amount of retries reached.")
```
Comment (ktdreyer) [inline mita/openstack.py:349]: Let's log the volume name here, `logger.error(f'Max retries for {volume.name} reached')`

File: mita/openstack.py [hunk 4]
```diff
@@ -376,14 +359,16 @@ def get_private_ip(self):
...
-        volumes = driver.list_volumes()
-        try:
-            return [v for v in volumes if v.name == name][0]
-        except IndexError:
-            raise RuntimeError("Unable to get volume")
+    def get_volume(self, name: str) -> StorageVolume:
+        """Retrieve the StorageVolume instance using the given name."""
+        url = f"/volumes?name={name}"
+        object_ = self.driver_v2.volumev2_connection.request(url).object
+
```
Comment (ktdreyer) [inline mita/openstack.py:366]: We have `object_.get("volumes", [])` here, and then later on we do `object_.get("volumes")[0]`...

Let's simply assign `volumes = object_.get("volumes", [])` once, and then use that `volumes` variable in the rest of the code.

File: mita/openstack.py [hunk 5]
```diff
@@ -376,14 +359,16 @@ def get_private_ip(self):
...
+        if len(object_.get("volumes", [])) != 1:
```
Comment (ktdreyer) [inline mita/openstack.py:367]: The old code did not care about multiple volumes with the same name, but here we're checking that `len()` is exactly one. Let's rewrite this to `if not volumes:`

File: mita/openstack.py [hunk 6]
```diff
@@ -376,14 +359,16 @@ def get_private_ip(self):
...
+            raise ResourceNotFound(f"Exact match failed for volume with name: {name}.")
+
+        volume_info = object_.get("volumes")[0]
```
Comment (ktdreyer) [inline mita/openstack.py:370]: After rewriting (per above) to use a new `volumes` variable, we can edit this line to simply be `volume_info = volumes[0]`

File: mita/openstack.py [hunk 7]
```diff
@@ -394,6 +379,8 @@ def create_node(self):
         if self.no_of_volumes:
             self._create_attach_volumes()
 
+        logger.info("Successfully created VM node with name %s.", self.node_name)
```
Comment (ktdreyer) [inline mita/openstack.py:382]: Let's take this out of this PR since it's logically separate

**Review by ktdreyer — CHANGES_REQUESTED** [review]

File: mita/openstack.py (hunk 4 above)
Comment (psathyan) [reply mita/openstack.py:367]: Yes it would make it more readable.

File: mita/openstack.py (hunk 6 above)
Comment (psathyan) [reply mita/openstack.py]: done

File: mita/openstack.py (hunk 5 above)
Comment (psathyan) [reply mita/openstack.py]: changed.

File: mita/openstack.py [hunk 8]
```diff
@@ -309,64 +309,50 @@ def _wait_until_ip_is_known(self):
 
     def _create_attach_volumes(self):
         """Create and attach the volumes."""
+        logger.info(
```
Comment (ktdreyer) [inline mita/openstack.py:312]: Fewer, clearer log statements :heart:

**Review by ktdreyer — APPROVED** [review]

File: mita/openstack.py (hunk 8 above)
Comment (psathyan) [reply mita/openstack.py:312]: thank you

**Pull request merged**